src/libsyntax/ext/expand.rs

```diff
@@ -282,6 +282,24 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
                     let expansion = self.expand_invoc(invoc, ext);
                     self.collect_invocations(expansion, &[])
                 } else if let InvocationKind::Attr { attr: None, traits, item } = invoc.kind {
+                    let derive_allowed = match item {
+                        Annotatable::Item(ref item) => match item.node {
+                            ast::ItemKind::Struct(..) |
+                            ast::ItemKind::Enum(..) |
+                            ast::ItemKind::Union(..) => true,
+                            _ => false,
+                        },
+                        _ => false,
+                    };
+                    if !derive_allowed {
+                        let span = item.attrs().iter()
+                            .find(|attr| attr.check_name("derive"))
+                            .expect("`derive` attribute should exist").span;
+                        self.cx.span_err(span,
+                                         "`derive` may only be applied to structs, enums, \
```

Comment:
Can you keep the existing output and avoid the oxford comma? Not because I'm against it but I'd prefer to keep this change small (so it doesn't affect `compile-fail/deriving-non-type.rs` at all).

During the impl period I'm going to audit all the existing diagnostic output in order to identify current conventions to write them down. Doing this I fully expect to find inconsistent wording and will try to reach consensus for them.

Reply:
I trust that the community will affirm the correctness of the serial comma at that time.